File: typescript/src/2025/advent_2025_day_01.ts
import { readFileSync } from "fs";

const input: string[] = readFileSync("./src/2025/advent_2025_day_01.txt", "utf-8").trim().split('\n')

function modulo(a: number, b: number) {
    return ((a % b) + b) % b;
}

function part1(): number {
    let password = 0;
    let dial = 50;
    for (const rotation of input) {
        dial += (rotation[0] == 'R') ? parseInt(rotation.slice(1)) : -parseInt(rotation.slice(1));
        password += (modulo(dial, 100) === 0) ? 1 : 0;
    }
    return password;
}

function part2(): number {
    let password = 0;
    let dial = 50;
    for (const rotation of input) {
        let delta = (rotation[0] == 'R') ? parseInt(rotation.slice(1)) : -parseInt(rotation.slice(1));
        password += Math.abs(Math.floor((dial + delta) / 100));
        if (delta < 0) {
            password -= (dial === 0) ? 1 : 0
            password += (modulo(dial + delta, 100) === 0) ? 1 : 0;
        }
        dial = modulo(dial + delta, 100);
    }
    return password;
}

console.log("Advent of Code 2025 - Day 1");
console.log(`Part 1: ${part1()}`);
console.log(`Part 2: ${part2()}`);

File: python/2025/advent_2025_day_01.py
def part_one(input_arr: list[str]) -> int:
    password = 0
    dial = 50
    for rotation in input_arr:
        dial += int(rotation[1:]) if rotation[0] == "R" else -int(rotation[1:])
        password += 1 if dial % 100 == 0 else 0
    return password


def part_two(input_arr: list[str]) -> int:
    password = 0
    dial = 50
    for rotation in input_arr:
        delta = int(rotation[1:]) if rotation[0] == "R" else -int(rotation[1:])
        password += abs((dial + delta) // 100)
        if delta < 0:
            password -= 1 if dial == 0 else 0
            password += 1 if (dial + delta) % 100 == 0 else 0
        dial = (dial + delta) % 100
    return password


input_arr: list[str] = open("advent_2025_day_01.txt").read().splitlines()

print("Advent of Code 2025 - Day 01")
print(f"Part One: {part_one(input_arr)}")
print(f"Part Two: {part_two(input_arr)}")
